File: ch2/Decimal_TO_Binary/algorithm.py
def _decimal_to_binary(number,binary = ""):
    if number == 0:
        print(binary); return
    binary = str(number%2) + binary #binary = "1" + binary if number & 1 else "0" + binary  
    _decimal_to_binary(number//2, binary=binary)

# if I did return binary
# It returns None value which is really weird ... maybe because it is the top of the stack
# but this version is very much cleaner

_decimal_to_binary(108)


#######################################################################################
#######################################################################################
#   this version returns a value ... I just returns the value from the top of the stack 
#   until the end


def decimal_to_binary(number,binary = ""):
    value = None
    if number == 0:
        value = binary
        return value
    binary = str(number%2) + binary #binary = "1" + binary if number & 1 else "0" + binary  
    value = decimal_to_binary(number//2, binary=binary)
    return value

# if I did return binary
# It returns None value which is really weird ... maybe because it is the top of the stack

File: ch2/Decimal_TO_Binary/algorithm.cpp
#include <iostream>
#include <string>

void decimal_to_binary(int number, std::string binary = "") {
    if (number == 0) { std::cout << binary << std::endl; return;}       
    binary = number & 1? "1" + binary : "0" + binary;         // more concise way ```binary = std::to_string(number%2) + binary``` but i find the former more "understandable"
    decimal_to_binary(number/2,binary=binary);
}

int main() {
    decimal_to_binary(108);             // check the python implementation if you want the algorithm "returns" a value because
}                                       // if instead of cout binary I returned I got core dumped because you are missing with the stack ... check the python comments